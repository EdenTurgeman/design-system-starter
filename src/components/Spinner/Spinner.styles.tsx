import styled from 'styled-components';
import React from 'react';

interface ISpinnerStyleProps {
  size: 'l' | 'm' | 's' | 'xl' | 'xs';
  role?: string;
  testId: string;
}

const sizes: Record<string, number> = {
  xs: 10,
  s: 20,
  m: 30,
  l: 45,
  xl: 90,
};

export const SpinnerSvg = ({ color }: { color: string }) => (
  <svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'>
    <defs>
      <linearGradient x1='28.154%' y1='63.74%' x2='74.629%' y2='17.783%' id={`${color}LoadingIndicatorGradient`}>
        <stop stopColor={color} offset='0%' />
        <stop stopColor={color} stopOpacity='0' offset='100%' />
      </linearGradient>
    </defs>
    <g transform='translate(2)' fill='none' fillRule='evenodd'>
      <circle stroke={`url(#${color}LoadingIndicatorGradient)`} strokeWidth='4' cx='10' cy='12' r='10' />
      <path d='M10 2C4.477 2 0 6.477 0 12' stroke={color} strokeWidth='4' />
    </g>
  </svg>
);

export const SpinnerContainer = styled.div<ISpinnerStyleProps>`
  width: ${({ size }) => sizes[size] ?? size}px;
  height: ${({ size }) => sizes[size] ?? size}px;
  font-size: 14px;
  line-height: 20px;
  border-width: 0;
  border-style: solid;
  border-color: rgba(0, 0, 0, 0.1);
  animation-name: spin;
  animation-duration: 500ms;
  animation-timing-function: linear;
  animation-iteration-count: infinite;
  transform-origin: center center;

  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }
`;
