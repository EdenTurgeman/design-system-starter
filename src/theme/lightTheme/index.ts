export * from './colors';
export * from './lightTheme';
