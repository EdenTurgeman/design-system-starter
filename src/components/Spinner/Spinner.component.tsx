import type { FunctionComponent } from 'react';
import React, { useContext } from 'react';
import { ThemeContext } from 'styled-components';
import { get } from 'lodash';
import { SpinnerContainer, SpinnerSvg } from './Spinner.styles';
import { Color } from '../../theme';

export interface SpinnerProps {
  size?: 'l' | 'm' | 's' | 'xl' | 'xs';
  color?: Color;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  'data-testId'?: string;
}

export const Spinner: FunctionComponent<SpinnerProps> = ({ size = 'm', color, ...props }) => {
  const theme = useContext(ThemeContext);

  const spinnerColor = color ? get(theme.colors, color) ?? color : theme.spinner.color;

  return (
    <SpinnerContainer testId={props['data-testId'] ?? 'spinner'} role={'progressbar'} size={size} {...props}>
      <SpinnerSvg color={spinnerColor} />
    </SpinnerContainer>
  );
};

Spinner.displayName = 'Spinner';
