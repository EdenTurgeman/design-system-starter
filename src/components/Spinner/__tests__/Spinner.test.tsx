import React from 'react';
import { render } from '@testing-library/react';
import { Spinner } from '../Spinner.component';
import { lightTheme, ThemeProvider } from '../../../theme';

describe('Spinner', () => {
  test('should match snapshot for spinner', () => {
    const { container } = render(
      <ThemeProvider theme={lightTheme}>
        <Spinner />
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  test('should match snapshot for spinner sizes', () => {
    const { container } = render(
      <ThemeProvider theme={lightTheme}>
        <>
          <Spinner size={'xs'} />
          <Spinner size={'s'} />
          <Spinner size={'m'} />
          <Spinner size={'l'} />
          <Spinner size={'xl'} />
        </>
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
  });
  test('should match snapshot for spinner color', () => {
    const { container } = render(
      <ThemeProvider theme={lightTheme}>
        <>
          <Spinner color={'yellow.600'} />
          <Spinner color={'red.600'} />
          <Spinner color={'green.600'} />
        </>
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
