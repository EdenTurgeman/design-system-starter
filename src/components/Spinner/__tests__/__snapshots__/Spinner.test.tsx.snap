// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner should match snapshot for spinner 1`] = `
.c0 {
  width: 30px;
  height: 30px;
  font-size: 14px;
  line-height: 20px;
  border-width: 0;
  border-style: solid;
  border-color: rgba(0,0,0,0.1);
  -webkit-animation-name: spin;
  animation-name: spin;
  -webkit-animation-duration: 500ms;
  animation-duration: 500ms;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

<div>
  <div
    class="c0"
    role="progressbar"
  >
    <svg
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <lineargradient
          id="#5C41E4LoadingIndicatorGradient"
          x1="28.154%"
          x2="74.629%"
          y1="63.74%"
          y2="17.783%"
        >
          <stop
            offset="0%"
            stop-color="#5C41E4"
          />
          <stop
            offset="100%"
            stop-color="#5C41E4"
            stop-opacity="0"
          />
        </lineargradient>
      </defs>
      <g
        fill="none"
        fill-rule="evenodd"
        transform="translate(2)"
      >
        <circle
          cx="10"
          cy="12"
          r="10"
          stroke="url(##5C41E4LoadingIndicatorGradient)"
          stroke-width="4"
        />
        <path
          d="M10 2C4.477 2 0 6.477 0 12"
          stroke="#5C41E4"
          stroke-width="4"
        />
      </g>
    </svg>
  </div>
</div>
`;

exports[`Spinner should match snapshot for spinner color 1`] = `
.c0 {
  width: 30px;
  height: 30px;
  font-size: 14px;
  line-height: 20px;
  border-width: 0;
  border-style: solid;
  border-color: rgba(0,0,0,0.1);
  -webkit-animation-name: spin;
  animation-name: spin;
  -webkit-animation-duration: 500ms;
  animation-duration: 500ms;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

<div>
  <div
    class="c0"
    role="progressbar"
  >
    <svg
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <lineargradient
          id="#f7c15cLoadingIndicatorGradient"
          x1="28.154%"
          x2="74.629%"
          y1="63.74%"
          y2="17.783%"
        >
          <stop
            offset="0%"
            stop-color="#f7c15c"
          />
          <stop
            offset="100%"
            stop-color="#f7c15c"
            stop-opacity="0"
          />
        </lineargradient>
      </defs>
      <g
        fill="none"
        fill-rule="evenodd"
        transform="translate(2)"
      >
        <circle
          cx="10"
          cy="12"
          r="10"
          stroke="url(##f7c15cLoadingIndicatorGradient)"
          stroke-width="4"
        />
        <path
          d="M10 2C4.477 2 0 6.477 0 12"
          stroke="#f7c15c"
          stroke-width="4"
        />
      </g>
    </svg>
  </div>
  <div
    class="c0"
    role="progressbar"
  >
    <svg
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <lineargradient
          id="#e95d39LoadingIndicatorGradient"
          x1="28.154%"
          x2="74.629%"
          y1="63.74%"
          y2="17.783%"
        >
          <stop
            offset="0%"
            stop-color="#e95d39"
          />
          <stop
            offset="100%"
            stop-color="#e95d39"
            stop-opacity="0"
          />
        </lineargradient>
      </defs>
      <g
        fill="none"
        fill-rule="evenodd"
        transform="translate(2)"
      >
        <circle
          cx="10"
          cy="12"
          r="10"
          stroke="url(##e95d39LoadingIndicatorGradient)"
          stroke-width="4"
        />
        <path
          d="M10 2C4.477 2 0 6.477 0 12"
          stroke="#e95d39"
          stroke-width="4"
        />
      </g>
    </svg>
  </div>
  <div
    class="c0"
    role="progressbar"
  >
    <svg
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <lineargradient
          id="#73b976LoadingIndicatorGradient"
          x1="28.154%"
          x2="74.629%"
          y1="63.74%"
          y2="17.783%"
        >
          <stop
            offset="0%"
            stop-color="#73b976"
          />
          <stop
            offset="100%"
            stop-color="#73b976"
            stop-opacity="0"
          />
        </lineargradient>
      </defs>
      <g
        fill="none"
        fill-rule="evenodd"
        transform="translate(2)"
      >
        <circle
          cx="10"
          cy="12"
          r="10"
          stroke="url(##73b976LoadingIndicatorGradient)"
          stroke-width="4"
        />
        <path
          d="M10 2C4.477 2 0 6.477 0 12"
          stroke="#73b976"
          stroke-width="4"
        />
      </g>
    </svg>
  </div>
</div>
`;

exports[`Spinner should match snapshot for spinner sizes 1`] = `
.c0 {
  width: 10px;
  height: 10px;
  font-size: 14px;
  line-height: 20px;
  border-width: 0;
  border-style: solid;
  border-color: rgba(0,0,0,0.1);
  -webkit-animation-name: spin;
  animation-name: spin;
  -webkit-animation-duration: 500ms;
  animation-duration: 500ms;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c1 {
  width: 20px;
  height: 20px;
  font-size: 14px;
  line-height: 20px;
  border-width: 0;
  border-style: solid;
  border-color: rgba(0,0,0,0.1);
  -webkit-animation-name: spin;
  animation-name: spin;
  -webkit-animation-duration: 500ms;
  animation-duration: 500ms;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c2 {
  width: 30px;
  height: 30px;
  font-size: 14px;
  line-height: 20px;
  border-width: 0;
  border-style: solid;
  border-color: rgba(0,0,0,0.1);
  -webkit-animation-name: spin;
  animation-name: spin;
  -webkit-animation-duration: 500ms;
  animation-duration: 500ms;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c3 {
  width: 45px;
  height: 45px;
  font-size: 14px;
  line-height: 20px;
  border-width: 0;
  border-style: solid;
  border-color: rgba(0,0,0,0.1);
  -webkit-animation-name: spin;
  animation-name: spin;
  -webkit-animation-duration: 500ms;
  animation-duration: 500ms;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c4 {
  width: 90px;
  height: 90px;
  font-size: 14px;
  line-height: 20px;
  border-width: 0;
  border-style: solid;
  border-color: rgba(0,0,0,0.1);
  -webkit-animation-name: spin;
  animation-name: spin;
  -webkit-animation-duration: 500ms;
  animation-duration: 500ms;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

<div>
  <div
    class="c0"
    role="progressbar"
  >
    <svg
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <lineargradient
          id="#5C41E4LoadingIndicatorGradient"
          x1="28.154%"
          x2="74.629%"
          y1="63.74%"
          y2="17.783%"
        >
          <stop
            offset="0%"
            stop-color="#5C41E4"
          />
          <stop
            offset="100%"
            stop-color="#5C41E4"
            stop-opacity="0"
          />
        </lineargradient>
      </defs>
      <g
        fill="none"
        fill-rule="evenodd"
        transform="translate(2)"
      >
        <circle
          cx="10"
          cy="12"
          r="10"
          stroke="url(##5C41E4LoadingIndicatorGradient)"
          stroke-width="4"
        />
        <path
          d="M10 2C4.477 2 0 6.477 0 12"
          stroke="#5C41E4"
          stroke-width="4"
        />
      </g>
    </svg>
  </div>
  <div
    class="c1"
    role="progressbar"
  >
    <svg
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <lineargradient
          id="#5C41E4LoadingIndicatorGradient"
          x1="28.154%"
          x2="74.629%"
          y1="63.74%"
          y2="17.783%"
        >
          <stop
            offset="0%"
            stop-color="#5C41E4"
          />
          <stop
            offset="100%"
            stop-color="#5C41E4"
            stop-opacity="0"
          />
        </lineargradient>
      </defs>
      <g
        fill="none"
        fill-rule="evenodd"
        transform="translate(2)"
      >
        <circle
          cx="10"
          cy="12"
          r="10"
          stroke="url(##5C41E4LoadingIndicatorGradient)"
          stroke-width="4"
        />
        <path
          d="M10 2C4.477 2 0 6.477 0 12"
          stroke="#5C41E4"
          stroke-width="4"
        />
      </g>
    </svg>
  </div>
  <div
    class="c2"
    role="progressbar"
  >
    <svg
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <lineargradient
          id="#5C41E4LoadingIndicatorGradient"
          x1="28.154%"
          x2="74.629%"
          y1="63.74%"
          y2="17.783%"
        >
          <stop
            offset="0%"
            stop-color="#5C41E4"
          />
          <stop
            offset="100%"
            stop-color="#5C41E4"
            stop-opacity="0"
          />
        </lineargradient>
      </defs>
      <g
        fill="none"
        fill-rule="evenodd"
        transform="translate(2)"
      >
        <circle
          cx="10"
          cy="12"
          r="10"
          stroke="url(##5C41E4LoadingIndicatorGradient)"
          stroke-width="4"
        />
        <path
          d="M10 2C4.477 2 0 6.477 0 12"
          stroke="#5C41E4"
          stroke-width="4"
        />
      </g>
    </svg>
  </div>
  <div
    class="c3"
    role="progressbar"
  >
    <svg
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <lineargradient
          id="#5C41E4LoadingIndicatorGradient"
          x1="28.154%"
          x2="74.629%"
          y1="63.74%"
          y2="17.783%"
        >
          <stop
            offset="0%"
            stop-color="#5C41E4"
          />
          <stop
            offset="100%"
            stop-color="#5C41E4"
            stop-opacity="0"
          />
        </lineargradient>
      </defs>
      <g
        fill="none"
        fill-rule="evenodd"
        transform="translate(2)"
      >
        <circle
          cx="10"
          cy="12"
          r="10"
          stroke="url(##5C41E4LoadingIndicatorGradient)"
          stroke-width="4"
        />
        <path
          d="M10 2C4.477 2 0 6.477 0 12"
          stroke="#5C41E4"
          stroke-width="4"
        />
      </g>
    </svg>
  </div>
  <div
    class="c4"
    role="progressbar"
  >
    <svg
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <lineargradient
          id="#5C41E4LoadingIndicatorGradient"
          x1="28.154%"
          x2="74.629%"
          y1="63.74%"
          y2="17.783%"
        >
          <stop
            offset="0%"
            stop-color="#5C41E4"
          />
          <stop
            offset="100%"
            stop-color="#5C41E4"
            stop-opacity="0"
          />
        </lineargradient>
      </defs>
      <g
        fill="none"
        fill-rule="evenodd"
        transform="translate(2)"
      >
        <circle
          cx="10"
          cy="12"
          r="10"
          stroke="url(##5C41E4LoadingIndicatorGradient)"
          stroke-width="4"
        />
        <path
          d="M10 2C4.477 2 0 6.477 0 12"
          stroke="#5C41E4"
          stroke-width="4"
        />
      </g>
    </svg>
  </div>
</div>
`;
